shell/platform/common/cpp/client_wrapper/include/flutter/encodable_value.h

```diff
@@ -6,16 +6,150 @@
+  using super = std::variant<std::monostate,
+                             bool,
+                             int32_t,
+                             int64_t,
+                             double,
+                             std::string,
+                             std::vector<uint8_t>,
+                             std::vector<int32_t>,
+                             std::vector<int64_t>,
+                             std::vector<double>,
+                             EncodableList,
+                             EncodableMap>;
+  using super::super;
+  using super::operator=;
+
+  explicit EncodableValue() : super() {}
+
+  // Avoid the C++17 pitfall of conversion from char* to bool. Should not be
+  // needed for C++20.
+  explicit EncodableValue(const char* string) : super(std::string(string)) {}
+  EncodableValue& operator=(const char* other) {
+    *this = std::string(other);
+    return *this;
+  }
+
+  // Override the conversion constructors from std::variant to make them
+  // explicit, to avoid implicit conversion.
+  //
+  // While implicit conversion can be convenient in some cases, it can have very
+  // surprising effects. E.g., calling a function that takes an EncodableValue
+  // but accidentally passing an EncodableValue* would, instead of failing to
+  // compile, go through a pointer->bool->EncodableValue(bool) chain and
+  // silently call the function with a temp-constructed EncodableValue(true).
+  template <class T>
+  constexpr explicit EncodableValue(T&& t) noexcept : super(t) {}
+
+  // Returns true if the value is null. Convenience wrapper since unlike the
+  // other types, std::monostate uses aren't self-documenting.
+  bool IsNull() const { return index() == 0; }
+
+  // Convience method to simplify handling objects received from Flutter where
+  // the values may be larger than 32-bit, since they have the same type on the
+  // Dart side, but will be either 32-bit or 64-bit here depending on the value.
+  //
+  // It is a programming error to call this method if the value doesn't contain
+  // either an int32_t or an int64_t.
+  int64_t LongValue() {
+    if (std::holds_alternative<int32_t>(*this)) {
+      return std::get<int32_t>(*this);
+    }
+    return std::get<int64_t>(*this);
+  }
+
+  // Convenience wrappers for nested collection types, to allow readable
+  // chaining when not doing type validation, keeping the [] lookup next to the
+  // type extraction. E.g.:
+  //   EncodableValue inner_value =
+  //       value.ListValue()[0].MapValue()[EncodableValue("some_key")];
+  // rather than
+  //   EncodableValue inner_value = std::get<EncodableMap>(
+  //      std::get<EncodableList>(value)[0])[EncodableValue("some_key")];
+  //
+  // It is a programming error to call these method if the value doesn't
+  // contain the correct collection type.
+  const EncodableList& ListValue() const {
+    return std::get<EncodableList>(*this);
+  }
+  EncodableList& ListValue() { return std::get<EncodableList>(*this); }
```

Comment:
I realize you said it was programmer error if the types don't match and while this is certainly true in Google code, in non-Google standard C++, I believe an exception will be thrown instead. Are these wrappers meant to be exception aware?

Reply:
I'll change it to say that it's an exception. (I'd still argue that's a programming error, but better to be precise)